mypy/test/data/check-inference.test

```diff
@@ -690,6 +690,60 @@ l = lb # E: Incompatible types in assignment (expression has type List[bool], va
 [builtins fixtures/for.py]
 
 
+-- Generic function inference with unions
+-- --------------------------------------
+
+
+[case testUnionInference]
+from typing import TypeVar, Union, List
+T = TypeVar('T')
+U = TypeVar('U')
+def f(x: Union[T, int], y: T) -> T: pass
+f(1, 'a')() # E: "str" not callable
+f('a', 1)() # E: "object" not callable
+
+def g(x: Union[T, List[T]]) -> List[T]: pass
+def h(x: List[str]) -> None: pass
+g('a')() # E: List[str] not callable
+g(['a']) # E: Argument 1 to "g" has incompatible type List[str]; expected "Union[None, List[None]]"
```

Comment:
This is an edge case where it's ambiguous whether we should pick `T = List[str]` or `T = str`. Maybe mypy should accept the call since it doesn't really matter which `T` is picked here, but I don't know how to achieve that. When generating constraints from the function call argument types, we can't generate either `T :> List[str]` or `T :> str` because there could be a second argument that forces the other choice. So the solver doesn't see any constraints and picks `T = None`, resulting in this error.

It would be better to at least give the error about not being able to infer the function type argument.